nghttp2: 1.19.0 -> 1.20.0

c0bw3b opened the pull request:
###### Motivation for this change
Package update. This PR is targeted at staging because a mass-rebuild is expected.

Full changelog : https://nghttp2.org/blog/2017/02/26/nghttp2-v1-20-0/

Corrected c-ares attribute name
Patching shebang in fetch-ocsp-response [REMOVED]
Adding SystemD support for nghttpx [REMOVED]


###### Things done

- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

c0bw3b committed:
nghttp2: 1.19.0 -> 1.20.0

Corrected c-ares attribute name

Full changelog : https://nghttp2.org/blog/2017/02/26/nghttp2-v1-20-0/

Mic92 commented on pkgs/development/libraries/nghttp2/default.nix:
```diff
@@ -17,27 +17,37 @@ with { inherit (stdenv.lib) optional; };
+    ++ optional enableAsioLib "--enable-asio-lib --with-boost-libdir=${boost}/lib";
+
+  preInstall = ''
+    substituteInPlace script/fetch-ocsp-response --replace '#!/usr/bin/env' '#!/run/current-system/sw/bin/env'
```
what is the difference between the two?

kirelagin replied:
The difference is that the latter won’t work on non-nixos, right? I thought, we were not doing this kind of things.

copumpkin replied:
Yeah, please don't use global paths like that. There are very few guarantees about what will be in `/sw/bin/` anyway, and even NixOS has `/usr/bin/env`. If you must patch it to something else, make it use an explicit full path from `coreutils`.

c0bw3b replied:
Yes my bad, this is not portable. I will correct.

Mic92 commented on pkgs/development/libraries/nghttp2/default.nix:
```diff
@@ -17,27 +17,37 @@ with { inherit (stdenv.lib) optional; };
 
   nativeBuildInputs = [ pkgconfig ];
-  buildInputs = [ openssl libev zlib libcares ]
+  buildInputs = [ openssl libev zlib c-ares systemd.dev ]
```
Adding systemd, causes an infinite recursion.

Mic92 replied:
From my understanding systemd is only used for the server component, maybe this could be a different derivation.

kirelagin replied:
It is probably a good idea to avoid systemd as much as possible, as it is Linux-only.

copumpkin replied:
Or make it conditional on `stdenv.isLinux`

c0bw3b replied:
SystemD is useful for `nghttpx` binary to support being instrumented with a unit service, which would open the door to a future service module.
But I agree it should be conditional on `stdenv.isLinux`. I will correct that.

@Mic92 : having two derivation would not be much more practical and it was already reversed in the past by edolstra in favor of a single multi-output derivation.

Mic92 replied:
@c0bw3b conditional inclusion still leads to a dependency cycle: 

```
nix-shell -p nghttp2 --show-trace                                           
error: while evaluating the attribute ‘nativeBuildInputs’ of the derivation ‘shell’ at /etc/nixos/nixpkgs/pkgs/build-support/trivial-builders.nix:7:14:
while evaluating the attribute ‘nativeBuildInputs’ of the derivation ‘nghttp2-1.20.0’ at /etc/nixos/nixpkgs/pkgs/development/libraries/nghttp2/default.nix:19:3:
while evaluating the attribute ‘configureFlags’ of the derivation ‘systemd-232’ at /etc/nixos/nixpkgs/pkgs/os-specific/linux/systemd/default.nix:13:3:
while evaluating the attribute ‘nativeBuildInputs’ of the derivation ‘kbd-2.0.3’ at /etc/nixos/nixpkgs/pkgs/os-specific/linux/kbd/default.nix:4:3:
while evaluating the attribute ‘autoconf’ of the derivation ‘hook’ at /etc/nixos/nixpkgs/pkgs/build-support/trivial-builders.nix:7:14:
while evaluating the attribute ‘nativeBuildInputs’ of the derivation ‘autoconf-2.69’ at /etc/nixos/nixpkgs/pkgs/development/tools/misc/autoconf/default.nix:4:3:
while evaluating the attribute ‘src’ of the derivation ‘gnum4-1.4.18’ at /etc/nixos/nixpkgs/pkgs/development/tools/misc/gnum4/default.nix:4:3:
while evaluating the attribute ‘nativeBuildInputs’ of the derivation ‘m4-1.4.18.tar.bz2’ at /etc/nixos/nixpkgs/pkgs/build-support/fetchurl/default.nix:96:3:
while evaluating the attribute ‘propagatedNativeBuildInputs’ of the derivation ‘curl-7.53.1’ at /etc/nixos/nixpkgs/pkgs/tools/networking/curl/default.nix:24:3:
infinite recursion encountered, at undefined position
```

c0bw3b commented:
Removed entirely the `systemd` dependency. It would require splitting _libnghttp2_ and _nghttp2_ binaries into two separate derivations to be able to build `nghttpx` with systemd support.

Also removed the shebang correction in `fetch-ocsp-response` because it does not prevent this python script to run and there is no need to add a dependency to `coreutils` for that.
